### PR TITLE
[Bugfix 1999] Correct attributes

### LINKED ISSUE #1999: Consequences of removing DiagonalOperation
I have two questions on this:

https://github.com/PennyLaneAI/pennylane/blob/927c4c67a1e1f9c79a5d7d4e89c496fac418657f/pennylane/ops/qubit/attributes.py#L169-L187

1) `IsingZZ` seems to be missing in this list, which maybe shows that the old model of observables being able to declare themselves that they are diagonal was superior over having a centralized list?

2) How can custom `Operation`s declare that they are diagonal now? Should they modify the `pennylane.ops.qubit.attributes.diagonal_in_z_basis`?

### PR DESCRIPTION
**Pull request by mariaschuld**
**Context:**

It was noted that the existing attributes miss some gates and, in one case, contain a wrong gate.

Adding the `IsingZZ` to `is_diagonal_in_z_basis` uncovered an issue: Since it does not have `_eigvals` defined, the eigenvalues are computed from the matrix using a numpy module, which means they are not differentiable in a backpropagation pipeline. 

I think this is an issue with inferring the eigenvalues from a matrix in a non-diffable manner, which goes against the grain of PL. For now we fix it by warning of this issue.

**Description of the Change:**

Remove/add gates. Add `_eigvals` method to `IsingZZ` and explain the `_eigvals` requirement in the attributes docstring.

**Benefits:** Fixes #1999

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -346,6 +346,14 @@
   ...     ...
   ```
 
+* The `IsingZZ` gate was added to the `diagonal_in_z_basis` attribute. For this 
+  an explicit `_eigvals` method was added.
+  [(#XXX)](https://github.com/PennyLaneAI/pennylane/pull/XXX)
```
Comment (mariaschuld) [inline doc/releases/changelog-dev.md:351]:
```suggestion
  [(#2113)](https://github.com/PennyLaneAI/pennylane/pull/2113)
```

File: doc/releases/changelog-dev.md [hunk 2]
```diff
@@ -346,6 +346,14 @@
...
+  
+* The `IsingXX`, `IsingYY` and `IsingZZ` gates were added to 
+  the `composable_rotations` attribute. 
+  [(#XXX)](https://github.com/PennyLaneAI/pennylane/pull/XXX)
```
Comment (mariaschuld) [inline doc/releases/changelog-dev.md:355]:
```suggestion
  [(#2113)](https://github.com/PennyLaneAI/pennylane/pull/2113)
```

File: doc/releases/changelog-dev.md [hunk 3]
```diff
@@ -442,6 +450,10 @@
   vector.
   [(#2028)](https://github.com/PennyLaneAI/pennylane/pull/2028)
 
+* The `Rot`, gate was removed from the `composable_rotations` attribute, since it 
+  does not fulfil this property. 
+  [(#XXX)](https://github.com/PennyLaneAI/pennylane/pull/XXX)
```
Comment (mariaschuld) [inline doc/releases/changelog-dev.md:455]:
```suggestion
  [(#2113)](https://github.com/PennyLaneAI/pennylane/pull/2113)
```

File: pennylane/ops/qubit/attributes.py [hunk 4]
```diff
@@ -178,10 +180,16 @@ def __contains__(self, obj):
...
     ]
 )
 """Attribute: Operations that are diagonal in the computational basis.
 
 For such operations, the eigenvalues provide all necessary information to
 construct the matrix representation in the computational basis.
+
+Note: Currently, the gates with this attribute need 
+to overwrite the eigenvalue method for differentiation to work. 
+The reason is that if this method missing, eigenvalues are computed from the matrix
```
Comment (mariaschuld) [inline pennylane/ops/qubit/attributes.py:193]:
```suggestion
The reason is that if this method is missing, eigenvalues are computed from the matrix
```

File: pennylane/ops/qubit/attributes.py [hunk 5]
```diff
@@ -178,10 +180,16 @@ def __contains__(self, obj):
...
+        "IsingZZ",
     ]
 )
 """Attribute: Operations that are diagonal in the computational basis.
 
 For such operations, the eigenvalues provide all necessary information to
 construct the matrix representation in the computational basis.
+
+Note: Currently, the gates with this attribute need 
+to overwrite the eigenvalue method for differentiation to work.
```
Comment (mariaschuld) [inline pennylane/ops/qubit/attributes.py:192]:
```suggestion
to explicitly define an eigenvalue representation for differentiation to work. 
```

File: pennylane/ops/qubit/attributes.py [hunk 6]
```diff
@@ -178,10 +180,16 @@ def __contains__(self, obj):
...
 )
 """Attribute: Operations that are diagonal in the computational basis.
 
 For such operations, the eigenvalues provide all necessary information to
 construct the matrix representation in the computational basis.
+
+Note: Currently, the gates with this attribute need 
+to overwrite the eigenvalue method for differentiation to work. 
+The reason is that if this method missing, eigenvalues are computed from the matrix
+representation using ``np.linalg.eigvals``, which is not differentiable.
```
Comment (mariaschuld) [inline pennylane/ops/qubit/attributes.py:194]:
```suggestion
representation using ``np.linalg.eigvals``, which fails for some tensor types that the matrix 
may be cast in for backpropagation devices.
```

**Comment by cvjjm** [comment]
Nice work!

**Comment by josh146** [comment]
@mariaschuld I may hold off on reviewing until the tests are passing if that is okay!

**Comment by mariaschuld** [comment]
> @mariaschuld I may hold off on reviewing until the tests are passing if that is okay!

Sorry @josh146 I should not have tagged you before checking the tests, this 20 min wait is too much for my impatience :)

**Review by josh146 — APPROVED** [review]
Looks good to me, thanks @mariaschuld!

**Pull request merged**